text/ps-signatures/README.md

```diff
@@ -0,0 +1,144 @@
+- Another test that shows a more elaborate scenario where a signature is requested on a mix of known and committed messages and then the knowledge of that signature is proved in addition to revealing some messages is shown in test [test_scenario_1](https://github.com/hyperledger/ursa/pull/46/files#diff-3b79c2d1f6accb7035421822bd62d6c6R10).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation that the RFC is describing is the implementation of sections 6.1 and 6.2 of the [PS signatures paper](https://eprint.iacr.org/2015/525) and an attempt has been made to use the same symbol names in the code as used in the paper. Some implementation optimizations have been done like the use of multi-pairings or multi-exponentiations (both constant and variable time). The PoK protocol for Pedersen commitments used is described in this [RFC](https://github.com/hyperledger/ursa-rfcs/pull/10). PS signatures require type-3 pairings.
+
+The `Signature` implementation is based on section 6.1 of the paper and implements "A Multi-Message Protocol" with the enhancement being the support of signer-known messages also as explained below:
+- Say the verification key supports 5 messages, the relevant generators are Y<sub>1</sub>, Y<sub>2</sub>, Y<sub>3</sub>, Y<sub>4</sub>, Y<sub>5</sub>.
+- The holder wants to reveal only last 3 message to the issuer but commits to the first 2 messages in a commitment C = g<sup>t</sup>Y<sub>1</sub><sup>m1</sup>Y<sub>2</sub><sup>m2</sup> (t is the randomness and should be preserved for unblinding the sig) and sends C to the issuer (along with a proof of knowledge).
+- The issuer now creates D = C * Y<sub>3</sub><sup>m3</sup>Y<sub>4</sub><sup>m4</sup>Y<sub>5</sub><sup>m5</sup> where m3, m4 and m5 are known to him. The issuer will now choose a random u and create the signature (g<sup>u</sup>, (XD)<sup>u</sup>) as described in the paper.
+- The holder will unblind the signature as described in the paper.
+
+The `PoKOfSignature` is based on section 6.2 of the paper and the implementation follows the paper almost as-it-is. However, the product of multiple pairings is replaced with a multi-pairing for efficiency. The paper specifies the check as e(sigma<sub>1</sub>, X_tilde) * e(sigma<sub>1</sub>, Y_tilde<sub>1</sub>)<sup>m<sub>1</sub></sup>...e(sigma<sub>1</sub>, g_tilde)<sup>t</sup> == e(sigma<sub>2</sub>, g_tilde). This is transformed into the check e(sigma<sub>1</sub>, X_tilde * Y_tilde<sub>1</sub><sup>m<sub>1</sub></sup> * ...g_tilde<sup>t</sup>) == e(sigma<sub>2</sub>, g_tilde). This check is transformed to e(sigma<sub>1</sub>, X_tilde * Y_tilde<sub>1</sub><sup>m<sub>1</sub></sup> * ...g_tilde<sup>t</sup>) * e(sigma<sub>2</sub>, g_tilde)<sup>-1</sup> == 1. Now this can be combined into one multi-pairing. The proof of knowledge (symbol pie in the paper) is done using [this protocol](https://github.com/hyperledger/ursa/pull/46/files#diff-168ead191b84bf0856f3d153e9da3ba5R14). 
```

Comment:
@hartm @khovratovich @manudrijvers @dubovitskaya This is an implementation level optimization which was not in the paper so i wanted to check with you guys that it is fine. @hartm reviewed it and found it ok.